### PR TITLE
Api

### PR DESCRIPTION
**Pull request by NikolajBjorner**

### CODE REVIEW

File: src/math/lp/column.h [hunk 1]
```diff
@@ -37,17 +37,21 @@ inline std::ostream& operator<<(std::ostream& out, lconstraint_kind k) {
...
     return a.second < b.second;
 }
-
-class ul_pair {
+class lar_term; // forward definition
+class column {
     u_dependency* m_lower_bound_witness = nullptr;
     u_dependency* m_upper_bound_witness = nullptr;
     bool          m_associated_with_row = false;  
+    lpvar         m_j; //the column index
```
Comment (NikolajBjorner) [inline src/math/lp/column.h:48]: the m_columns array stores column objects. The index m_j is always going to be the same as the index where the column is stored.
Do you evern need to access m_j after having lost track of how you accessed m_columns?

File: src/math/lp/column.h [hunk 2]
```diff
@@ -37,17 +37,21 @@ inline std::ostream& operator<<(std::ostream& out, lconstraint_kind k) {
...
+inline bool compare(const std::pair<mpq, lpvar> & a, const std::pair<mpq, lpvar> & b) {
     return a.second < b.second;
 }
-
-class ul_pair {
+class lar_term; // forward definition
+class column {
     u_dependency* m_lower_bound_witness = nullptr;
     u_dependency* m_upper_bound_witness = nullptr;
     bool          m_associated_with_row = false;
```
Comment (NikolajBjorner) [inline src/math/lp/column.h:47]: currently, m_associated_with_row is set in tandem with m_term, so redundant?

File: src/math/lp/column.h [hunk 3]
```diff
@@ -37,17 +37,21 @@ inline std::ostream& operator<<(std::ostream& out, lconstraint_kind k) {
...
     u_dependency* m_lower_bound_witness = nullptr;
     u_dependency* m_upper_bound_witness = nullptr;
     bool          m_associated_with_row = false;  
+    lpvar         m_j; //the column index
+    lar_term*     m_term = nullptr;
 public:
-    // TODO - seems more straight-forward to just expose ul_pair as a struct with direct access to attributes.
-    
+    lar_term*  term() const { return m_term; }
+    lar_term*& term() { return m_term; } // for setting m_term
```
Comment (NikolajBjorner) [inline src/math/lp/column.h:52]: you actually use this?

File: src/math/lp/lar_solver.cpp [hunk 4]
```diff
@@ -1466,64 +1387,41 @@ namespace lp {
...
-
     struct lar_solver::undo_add_column : public trail {
         lar_solver& s;
         undo_add_column(lar_solver& s) : s(s) {}
         void undo() override {
+            auto& col = s.m_columns.back();
+            if (col.term() != nullptr) {
+                if (s.m_need_register_terms)
+                    s.deregister_normalized_term(*col.term());
+                delete col.term();
```
Comment (NikolajBjorner) [inline src/math/lp/lar_solver.cpp:1404]: or delete s.m_terms.back();

File: src/math/lp/column.h (hunk 1 above)
Comment (levnach) [reply src/math/lp/column.h:48]: Addressed

File: src/math/lp/column.h (hunk 3 above)
Comment (levnach) [reply src/math/lp/column.h]: No, it is not used. Removed now.

File: src/math/lp/lar_solver.cpp (hunk 4 above)
Comment (levnach) [reply src/math/lp/lar_solver.cpp:1404]: we have col == s.m_terms.back() here already. I am not sure what we gain by using "delete s.m_terms.back()".

File: src/sat/smt/arith_sls.cpp [hunk 5]
```diff
@@ -59,18 +59,18 @@ namespace arith {
...
             }
-            if (v == 52) {
-                verbose_stream() << "update v" << v << " := " << val << "\n";
-                for (lp::lar_term::ival const& arg : term) {
-                    auto t2 = s.lp().column2tv(arg.column());
-                    auto w = s.lp().local_to_external(t2.id());
-                    verbose_stream() << "v" << w << " := " << m_vars[w].m_best_value << " * " << to_numeral(arg.coeff()) << "\n";
-                }
-            }
+            // if (v == 52) {
```
[the remark this answers is not in the document]
Comment (levnach) [reply src/sat/smt/arith_sls.cpp]: removed

**Pull request merged**